[1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130760
* #131408
* #132391
* #130239
* __->__ #126697

**Summary**
This PR adds a new private placement type `_StridedShard` for FSDP2 + TP style tensor sharding. The previously used `Shard` placement type cannot produce correct `full_tensor()` result because it assumes the tensor to be first sharded over `dp` mesh dimension then `tp` mesh dimension which does not hold true in FSDP2 + TP case.

**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

XilunWu committed:
[dtensor] introduce StridedShard placement type

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] introduce StridedShard placement type"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] introduce StridedShard placement type"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][dtensor] introduce StridedShard placement type"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -224,3 +226,14 @@ def compute_local_stride(
     )
+
+
+def compute_shard_coordinate(
```
this function seems not getting used?

wanchaol reviewed (commented):
Looks to be in the right direction, please see inlined comments.

XilunWu commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -340,6 +340,131 @@ def __str__(self) -> str:
 
 
+# kw_only is only available in python >= 3.10
+kw_only_dataclass = dict(kw_only=True) if "kw_only" in dataclass.__kwdefaults__ else {}
```
The main motivation is the idea of Strided Sharding is not very straightforward and sometimes hard to understand/remember. Adding a good descriptive name to the argument `split_factor` is a way to address this IMO.

XilunWu committed:
Update on "[WIP][1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[WIP][1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -340,6 +340,128 @@ def __str__(self) -> str:
+    dimension before processing the _StridedShard placement, as if the sharding happened
+    right-to-left. In the example above, the tensor should first be sharded on the "tp"
+    dimension into 2 shards before being sharded on the "dp" dimension. Therefore, the
+    `split_factor` of the _StridedShard placement on "dp" dim is 2.
```
I think we should add a TODO in this comment to:
1. try to see how to unify this with the `Shard` placement later
2. make strided sharding work with uneven sharding and padding

wanchaol commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -21,6 +21,7 @@ def compute_local_shape(
     Compute the shape of a local shard of the given DTensor on its current
     coordinate of the mesh.
     """
+    # TODO(xilun): address new Shard type
```
hmmm would local shape really matters for strided shard placement? I think for either shard/strideShard they should produce the same local shape, except for the case when it works with uneven sharding? Let's be more clear what the TODO mean or simply remove this

XilunWu replied:
You're right, strided sharding only differs on where the shard is placed and it doesn't affect shape but only offset. 

For uneven padding, I think the above claim still holds true because of SPMD -- every shard will have the same shape after padding.

wanchaol commented on torch/distributed/_tensor/_utils.py:
```diff
@@ -85,6 +86,7 @@ def compute_local_shape_and_global_offset(
     rank6 -- local_shape:[0,], global_offset:[2,]
     rank7 -- local_shape:[0,], global_offset:[2,]
     """
+    # TODO(xilun): address new Shard type
```
```suggestion
    # TODO(xilun): address strided sharding placement
```

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -340,6 +340,128 @@ def __str__(self) -> str:
+        assert tensor.size(self.dim) % total_split == 0, (
+            "_StridedShard currently only allows even sharding but got tensor size"
+            f" {tensor.size(self.dim)} on dim {self.dim} and total split"
+            f" {total_split}={num_chunks} * {self.split_factor}"
```
```suggestion
            f" {total_split}="
```

wanchaol reviewed (approved):
looks good!

XilunWu committed:
Update on "[1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Summary**
This PR adds a new private placement type `_StridedShard` for FSDP2 + TP style tensor sharding. The previously used `Shard` placement type cannot produce correct `full_tensor()` result because it assumes the tensor to be first sharded over `dp` mesh dimension then `tp` mesh dimension which does not hold true in FSDP2 + TP case.

**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu committed:
Update on "[1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Summary**
This PR adds a new private placement type `_StridedShard` for FSDP2 + TP style tensor sharding. The previously used `Shard` placement type cannot produce correct `full_tensor()` result because it assumes the tensor to be first sharded over `dp` mesh dimension then `tp` mesh dimension which does not hold true in FSDP2 + TP case.

**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606120).

XilunWu committed:
Update on "[1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Summary**
This PR adds a new private placement type `_StridedShard` for FSDP2 + TP style tensor sharding. The previously used `Shard` placement type cannot produce correct `full_tensor()` result because it assumes the tensor to be first sharded over `dp` mesh dimension then `tp` mesh dimension which does not hold true in FSDP2 + TP case.

**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse penguinwu tianyu-l yf225 chauhang

Differential Revision: [D60606120](https://our.internmc.facebook.com/intern/diff/D60606120)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60606120).

XilunWu committed:
Update on "[1/N][dtensor] introduce StridedShard placement type and _split_tensor() logic"


**Summary**
This PR adds a new private placement type `_StridedShard` for FSDP2 + TP style tensor sharding. The previously used `Shard` placement type cannot produce correct `full_tensor()` result because it assumes the tensor to be first sharded over `dp` mesh dimension then `tp` mesh dimension which does not hold true in FSDP2 + TP case.

**Test**
`pytest test/distributed/_tensor/test_utils.py -s -k strided_sharding`

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse penguinwu tianyu-l yf225 chauhang

Differential Revision: [D60606120](https://our.internmc.facebook.com/intern/diff/D60606120)

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60917088).
